### PR TITLE
Fix KeyError in ROS.19

### PR DESCRIPTION
**Pull request by derekgengenbacher-wf**
#### Reason for change
ROS.19 can fail with a KeyError

#### Description of change
Updated usages of `principal_currency_values`

#### Steps to Test
CI

**review**:
@Arelle/arelle

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: arelle/plugin/validate/ROS/rules/ros.py [hunk 1]
```diff
@@ -368,13 +368,13 @@ def rule_ros19(
         for unit, count in unit_counts.items():
             if count > principal_currency_value_count:
                 modelObjects = principal_currency_facts
-                message = _("'PrincipalCurrencyUsedInBusinessReport' has a value of %(principal_currency_value)s, "
+                message = _("'PrincipalCurrencyUsedInBusinessReport' has a value of %(principal_currency_values)s, "
```
Comment (brettkail-wk) [inline arelle/plugin/validate/ROS/rules/ros.py:371]: This output seems worse ("has a value of {USD}" vs "has a value of USD").  Why this change?

File: arelle/plugin/validate/ROS/rules/ros.py [hunk 2]
```diff
@@ -368,13 +368,13 @@ def rule_ros19(
...
                             "which must match the functional(majority) unit of the financial statement.")
                 continue
     if message:
         yield Validation.warning(
             "ROS.19",
             message,
             modelObject=modelObjects,
-            principal_currency_value=principal_currency_value
+            principal_currency_values=principal_currency_values if principal_currency_values else None
```
Comment (brettkail-wk) [inline arelle/plugin/validate/ROS/rules/ros.py:379]: Is the intent to turn empty set into `None`?  If yes, out of curiosity, why?  If no, why not simply `principal_currency_values=principal_currency_values`?

**Review by austinmatherne-wk — DISMISSED** [review]

File: arelle/plugin/validate/ROS/rules/ros.py [hunk 3]
```diff
@@ -376,5 +376,6 @@ def rule_ros19(
             "ROS.19",
             message,
             modelObject=modelObjects,
+            principal_currency_values=principal_currency_values,
             principal_currency_value=principal_currency_value
```
Comment (brettkail-wk) [inline arelle/plugin/validate/ROS/rules/ros.py:380]: I think the `principal_currency_value` line still needs to be removed (this was presumably the cause of the error since thatvar is only defined if there's exactly one).

Comment (derekgengenbacher-wf) [reply arelle/plugin/validate/ROS/rules/ros.py:380]: `principal_currency_value` will show up as a None if it is not set.

Comment (austinmatherne-wk) [reply arelle/plugin/validate/ROS/rules/ros.py:380]: It'll throw a 'principal_currency_value' is not defined exception if the conditional check on line 359 is True because it's only assigned in the else branch.

Comment (brettkail-wk) [reply arelle/plugin/validate/ROS/rules/ros.py:380]: Can you update the PR description with the error you're trying to fix?  I assumed it was `NameError: name 'principal_currency_value' is not defined` (and that `KeyError` in the summary was a mistake), but perhaps I misunderstand the problem.

Comment (austinmatherne-wk) [reply arelle/plugin/validate/ROS/rules/ros.py:380]: I haven't seen the original exception, but I believe it does indeed throw a KeyError because of how this gets passed through and handled by the logger (it looks for an arg matching the name of the variable within the message).

Comment (brettkail-wk) [reply arelle/plugin/validate/ROS/rules/ros.py:380]: The `principal_currency_value` isn't defined in the "then" branch, so shouldn't the code fail (before and after the PR) with `NameError` on the `principal_currency_value=principal_currency_value` line before `Validation.warning` fails to resolve `%(principal_currency_values)s`?  It seems like we're missing a test.

Comment (derekgengenbacher-wf) [reply arelle/plugin/validate/ROS/rules/ros.py]: I restructured the validation to yield at the time of validation rather than the end of the validation.

File: arelle/plugin/validate/ROS/rules/ros.py [hunk 4]
```diff
@@ -367,14 +371,10 @@ def rule_ros19(
...
-                            "which must match the functional(majority) unit of the financial statement.")
-                continue
-    if message:
-        yield Validation.warning(
-            "ROS.19",
-            message,
-            modelObject=modelObjects,
-            principal_currency_value=principal_currency_value
-        )
+                yield Validation.warning(
```
Comment (austinmatherne-wk) [inline arelle/plugin/validate/ROS/rules/ros.py:374]: Can this warning potentially fire multiple times within the containing loop now?

**Review by brettkail-wk — APPROVED** [review]
